File: Add Binary.java
public class Solution {
    public String addBinary(String a, String b) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(a.isEmpty()&&b.isEmpty())
        {
            return "0";
        }
        
        int addZeros = Math.abs(a.length() - b.length());
        char []zeros = new char[addZeros];
        Arrays.fill(zeros, '0');
        StringBuffer sb = new StringBuffer();
        sb.append(zeros);
        String formatedStrA = a;
        String formatedStrB = b;
        if(a.length() < b.length())
        {
            formatedStrA = sb.append(a).toString();
        }
        else
        {
            formatedStrB = sb.append(b).toString();
        }
        
        int carryOut = 0;
        StringBuffer ret = new StringBuffer();
        for(int i = formatedStrA.length()-1; i> -1; i--)
        {
            int val1 = formatedStrA.charAt(i) - '0';
            int val2 = formatedStrB.charAt(i) - '0';
            int val = (val1^val2)^carryOut;
            carryOut = (val1&val2)|(val1&carryOut)|(val2&carryOut);
            ret.insert(0, (char)(val+'0'));
        }
        if(carryOut==1)
        {
            ret.insert(0, carryOut);
        }
        
        return ret.toString();
    }
}

File: Add Binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(a.empty()&&b.empty())
        {
            return "0";
        }
        
        if(a.size()>b.size())
        {
            string zeros;
            zeros.append(a.size()-b.size(), '0');
            b = zeros+b;
        }
        else if(b.size()>a.size())
        {
            string zeros;
            zeros.append(b.size()-a.size(), '0');
            a = zeros+a;
        }
        
        int carryout = 0;
        int n = a.size();
        string result;
        for(int i = n-1; i> -1; i--)
        {
            int val1 = a[i]-'0';
            int val2 = b[i]-'0';
            int ret = val1^val2;
            ret = ret^carryout;
            carryout = (val1&val2)|(val2&carryout)|(val1&carryout);
            result.insert(result.begin(), ret+'0');
        }
        
        if(carryout>0)
        {
            result.insert(result.begin(), carryout+'0');
        }
        
        return result;
    }
};
